File: python_code/cache.py
import threading as th

import memory_primitives as mp


class Cache:
    def __init__(
        self,
        cache_size: int,
    ):
        self.cache_size = cache_size
        self.key_map = {i: None for i in range(cache_size)}
        self.cache = {i: None for i in range(cache_size)}
        self.locks = {i: th.Lock() for i in range(cache_size)}

    def read_no_sync(self, memory_address: int) -> None | mp.MemoryItem:
        key = memory_address % self.cache_size
        if self.key_map[key] != memory_address:
            return None
        return self.cache[key]

    def read(self, memory_address: int) -> None | mp.MemoryItem:
        with self.get_lock(memory_address):
            return self.read_no_sync(memory_address)
    
    def write(
            self,
            memory_address: int,
            data,
            status: str,
            wtag: int,
        ) -> None | mp.MemoryItem:
        with self.get_lock(memory_address):
            key = memory_address % self.cache_size
            if self.key_map[key] == memory_address:
                self.cache[key].data = data
                self.cache[key].status = status
                self.cache[key].wtag = wtag
                return self.cache[key]
            else:
                self.key_map[key] = memory_address
                self.cache[key] = mp.MemoryItem(
                    data=data,
                    status=status,
                    wtag=wtag,
                )
                return self.cache[key]
        
    def remove(self, memory_address: int) -> None:
        with self.get_lock(memory_address):
            key = memory_address % self.cache_size
            if self.key_map[key] == memory_address:
                self.key_map[key] = None
                self.cache[key] = None
        
    def get_lock(self, memory_address: int) -> th.Lock:
        key = memory_address % self.cache_size
        return self.locks[key]

File: java_code/edcs/src/main/edcs/project/Cache.java
package main.edcs.project;

import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

public class Cache {

    private int cacheSize;
    protected ConcurrentHashMap<Integer, Integer> keyMap;
    private ConcurrentHashMap<Integer, MemoryItem> cache;
    private ConcurrentHashMap<Integer, Lock> locks;

    public Cache(int cacheSize) {
        this.cacheSize = cacheSize;
        this.keyMap = new ConcurrentHashMap<>(cacheSize);
        this.cache = new ConcurrentHashMap<>(cacheSize);
        this.locks = new ConcurrentHashMap<>(cacheSize);

        for (int i = 0; i < cacheSize; i++) {
            keyMap.put(i, -1);
            cache.put(i, new MemoryItem(null, "C", -1));
            locks.put(i, new ReentrantLock());
        }
    }

    /*
    Description: Read from cache without synchronization.
    Used in the dumpcache functionality, which is called just for
    debugging/checking purposes.
     */
    public MemoryItem readNoSync(int memoryAddress) {
        int key = memoryAddress % cacheSize;
        if (memoryAddress != keyMap.get(key)) {
            return null;
        }
        return cache.get(key);
    }

    // Description: Read from cache with synchronization.
    public MemoryItem read(int memoryAddress) {
        synchronized (this.getLock(memoryAddress)) {
            return readNoSync(memoryAddress);
        }
    }

    // Description: Write to cache with synchronization.
    public MemoryItem write(int memoryAddress, Object data, String status, long wtag) {
        synchronized (this.getLock(memoryAddress)) {
            int key = memoryAddress % cacheSize;
            MemoryItem memoryItem = cache.get(key);
            if (memoryAddress == keyMap.get(key)) {
                memoryItem.setData(data);
                memoryItem.setStatus(status);
                memoryItem.setWtag(wtag);
            } else {
                keyMap.put(key, memoryAddress);
                memoryItem = new MemoryItem(data, status, wtag);
                cache.put(key, memoryItem);
            }
            return memoryItem;
        }
    }

    // Description: Remove an item from the cache with synchronization.
    public void remove(int memoryAddress) {
        synchronized (this.getLock(memoryAddress)) {
            int key = memoryAddress % cacheSize;
            if (memoryAddress == keyMap.get(key)) {
                keyMap.put(key, -1);
                cache.put(key, new MemoryItem(null, "C", -1));
            }
        }
    }

    // Description: Get the lock for a given memory address.
    public Lock getLock(int memoryAddress) {
        int key = memoryAddress % cacheSize;
        return locks.get(key);
    }
}